Add firewall to Pulumi

jemrobinson opened the pull request:
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the ~**develop**~ **python-migration** branch.
- [x] Your branch is up-to-date with the **target branch** (you probably started your branch from the target but it may have changed since then).
- [x] If-and-only-if your changes are not yet ready to merge, you have marked this pull request as a **draft** pull request and added '[WIP]' to the title.
- [x] If-and-only-if you have changed any Powershell code, you have run the code formatter. You can do this with `./tests/AutoFormat_Powershell.ps1 -TargetPath <path to file or directory>`.

### :arrow_heading_up: Summary

- Add Azure firewall and associated rules to Pulumi
- Add route table to redirect traffic to firewall
- Add private DNS zones to allow access to automation account

### :closed_umbrella: Related issues

Closes #1374 . Closes #1376. Closes #1384.

### :microscope: Tests

Tested on a local deployment

jemrobinson committed:
:sparkles: Add firewall to SHM

jemrobinson committed:
:sparkles: Add SHM route table

jemrobinson committed:
:closed_lock_with_key: Add domain admin password to backend keyvault

jemrobinson committed:
:sparkles: Add route table to direct traffic through the firewall. Add private links to automation account.

JimMadge commented on data_safe_haven/pulumi/components/shm_firewall.py:
```diff
@@ -0,0 +1,1128 @@
+        )
+
+        # Get firewall subnet
+        snet_firewall = network.get_subnet_output(
+            subnet_name=props.subnet_firewall_name,
+            resource_group_name=props.resource_group_name,
+            virtual_network_name=props.subnet_firewall_virtual_network_name,
+        )
```
This is saying, 'find me a subnet with this name, belonging to this virtual network, in this resource group'.

I assume this is a resource in the same stack?

We could pass the Python object about, but maybe this way keeps the code tidier when there are many resources?

jemrobinson replied:
Yes. You can either create a `Subnet` as its own resource or as a subresource inside a `VirtualNetwork`. In order to stop the subnets being destroyed and recreated on a re-run we currently do the second one, which means there isn't a `Subnet` object to pass around. I've thought of a way we might be able to do the first one though - if it works, I'll update this logic.

JimMadge replied:
I see, I think I might have come across this before too.

If I remember correctly the vnet object has an attribute containing its subnets (although that might be Terraform).

JimMadge replied:
I do that [here](https://github.com/alan-turing-institute/bureau/blob/main/build/bureaubuild/infrastructure.py)

```python
    vnet = network.VirtualNetwork(
        'vnet',
        virtual_network_name='vnet',
        resource_group_name=resource_group.name,
        address_space=network.AddressSpaceArgs(
            address_prefixes=['10.0.0.0/16'],
        ),
        subnets=[network.SubnetArgs(
            name='default',
            address_prefix='10.0.0.0/24'
        )]
    )
```

and later

```python
        'focal': BuildVMArgs(
            resource_group=resource_group,
            image_reference=compute.ImageReferenceArgs(
                offer='0001-com-ubuntu-server-focal',
                publisher='canonical',
                sku='20_04-lts-gen2',
                version='latest',
            ),
            subnet=vnet.subnets[0],
            private_key=private_key
        ),
```

then in `BuildVM.init`

```python
        nic = network.NetworkInterface(
            f'{name}_nic',
            resource_group_name=args.resource_group.name,
            location=args.resource_group.location,
            ip_configurations=[
                network.NetworkInterfaceIPConfigurationArgs(
                    name='buildvmipconfig',
                    subnet=network.SubnetArgs(id=args.subnet.id),
                    public_ip_address=network.PublicIPAddressArgs(
                        id=public_ip.id
                    )
                )
            ],
            opts=child_opts,
        )
```

So it looks like `vnet.subnets` is a list of `pulumi_azure_native.network.subnet` rather than `SubnetArgs`.

jemrobinson replied:
Yes, but they're ordered rather than named. I suppose we could match up the ordering the named return values?

JimMadge replied:
Oh, yes that does make it a bit tricker.

Could do something like

```python
def get_subnet(subnets: [subnet], name: str) -> Optional(subnet):
    for subnet in subnets:
        if subnet.name == name:
            return subnet
    return None
```

jemrobinson replied:
Done by running `get_subnet` and adding the `GetSubnetResult` as an output from the component.

JimMadge commented on data_safe_haven/pulumi/components/shm_firewall.py:
```diff
@@ -0,0 +1,1128 @@
+                            ],
+                            source_addresses=[
+                                "*"
+                                # "{{network.vnetRepositoriesTier2.subnets.mirrorsExternal.cidr}}",
+                                # "{{network.vnetRepositoriesTier2.subnets.proxies.cidr}}",
+                                # "{{network.vnetRepositoriesTier3.subnets.mirrorsExternal.cidr}}",
+                                # "{{network.vnetRepositoriesTier3.subnets.proxies.cidr}}"
```
These will exist eventually but not yet?

jemrobinson replied:
Yes - this is so we don't forget to add them once they exist :)

JimMadge commented on data_safe_haven/pulumi/components/shm_firewall.py:
```diff
@@ -0,0 +1,1128 @@
+        route = network.Route(
+            f"{self._name}_via_firewall",
+            address_prefix="0.0.0.0/0",
+            # next_hop_ip_address=(firewall.ip_configurations)[0].private_ip_address,
+            next_hop_ip_address=firewall.ip_configurations.apply(
+                lambda cfgs: cfgs[0].private_ip_address
+            ),
```
```suggestion
            next_hop_ip_address=firewall.ip_configurations.apply(
                lambda cfgs: cfgs[0].private_ip_address
            ),
```

jemrobinson committed:
:bug: Allow route table to ignore added routes, preventing them from being removed on redeploy

jemrobinson committed:
:rotating_light: Mark ResourceOptions as Optional

jemrobinson committed:
:sparkles: Add NAT rule to RDP to domain controller via firewall

jemrobinson committed:
:sparkles: Add subnet functionality to AzureIPv4Range

jemrobinson committed:
:rotating_light: Fix typing issues

jemrobinson committed:
:recycle: Move subnet definitions out of SHM virtual network

jemrobinson committed:
:rotating_light: Fix typing

jemrobinson committed:
:recycle: Standardise subnet handling across SHM and SRE

jemrobinson committed:
:bug: Allow PowershellDSC to reboot the VM during configuration

jemrobinson committed:
:sparkles: Add public DNS address for domain controller (to be replaced in future) and reboot DC during SHM deployment

jemrobinson committed:
:bug: Fix secret overwriting when value is unchanged

jemrobinson committed:
:recycle: Refactor backend resource names

jemrobinson committed:
:wrench: Allow raw.githubusercontent.com from domain controller for script download

jemrobinson committed:
:bug: Ensure that SHM DNS zone is delegated to SRE zone before trying to create SSL certificate

jemrobinson commented:
I think I've addressed all of @JimMadge's comments. I've checked that creation/teardown of SHM and SRE work as expected. This involved adding a few additional cross-referenced dependencies to ensure that resources are created in the correct order. There are also some general type-hinting fixes that I added while debugging.

JimMadge commented on data_safe_haven/commands/deploy_shm_command.py:
```diff
@@ -115,7 +121,7 @@ def handle(self) -> None:
         """Request any missing config values and add them to the config"""
         # Request FQDN if not provided
-        fqdn = self.option("fqdn")
+        fqdn = cast(str, self.option("fqdn"))
```
Does this just trick the type checker that the result of `self.option("fqdn")` is always a string?

What if it is empty?
What if it isn't a string?

jemrobinson replied:
This is a limitation in the type signature of `Command.option`. It returns `dict[Unknown, Unknown] | Unknown | Any | Literal[False]` which are all of the possibilities that `cleo` can coerce the command-line arguments into.

However, this isn't very useful for type checking. I'll have another think about this.

JimMadge commented on data_safe_haven/external/api/graph_api.py:
```diff
@@ -535,7 +540,7 @@ def delete_application(
             ) from exc
 
-    def get_application_by_name(self, application_name: str) -> JSONType:
+    def get_application_by_name(self, application_name: str) -> Dict[str, Any] | None:
```
```suggestion
    def get_application_by_name(self, application_name: str) -> Optional[Dict[str, Any]]:
```

JimMadge commented on data_safe_haven/external/api/graph_api.py:
```diff
@@ -555,13 +562,16 @@ def get_service_principal_by_name(self, service_principal_name: str) -> JSONType
             return None
 
-    def get_id_from_application_name(self, application_name: str) -> str:
+    def get_id_from_application_name(self, application_name: str) -> str | None:
```
```suggestion
    def get_id_from_application_name(self, application_name: str) -> Optional[str]:
```

JimMadge commented on data_safe_haven/external/api/graph_api.py:
```diff
@@ -555,13 +562,16 @@ def get_service_principal_by_name(self, service_principal_name: str) -> JSONType
             return None
 
-    def get_id_from_groupname(self, group_name: str) -> str:
+    def get_id_from_groupname(self, group_name: str) -> str | None:
```
```suggestion
    def get_id_from_groupname(self, group_name: str) -> Optional[str]:
```

JimMadge commented on data_safe_haven/external/api/graph_api.py:
```diff
@@ -571,7 +581,7 @@ def get_id_from_groupname(self, group_name: str) -> str:
             return None
 
-    def get_id_from_username(self, username: str) -> str:
+    def get_id_from_username(self, username: str) -> str | None:
```
```suggestion
    def get_id_from_username(self, username: str) -> Optional[str]:
```

JimMadge commented on data_safe_haven/external/api/graph_api.py:
```diff
@@ -664,7 +686,7 @@ def http_post(self, url: str, **kwargs: Any) -> requests.Response:
             ) from exc
 
-    def read_applications(self) -> JSONType:
+    def read_applications(self) -> Sequence[Dict[str, Any]]:
```
Do we really not know what concrete collection is returned here?

JimMadge commented on data_safe_haven/external/interface/azure_ipv4_range.py:
```diff
@@ -1,24 +1,56 @@
         """Azure reserves x.x.x.1 for the default gateway and (x.x.x.2, x.x.x.3) to map Azure DNS IPs."""
         return list(self.hosts())[3:]
+
+    def next_subnet(self, number_of_addresses: int) -> "AzureIPv4Range":
+        """Find the next unused subnet of a given size"""
+        if not math.log2(number_of_addresses).is_integer():
+            raise DataSafeHavenIPRangeException(
+                f"Number of address '{number_of_addresses}' must be a power of 2"
+            )
+        ip_address_first = self[0]
+        while True:
+            ip_address_last = ip_address_first + int(number_of_addresses - 1)
+            with suppress(DataSafeHavenIPRangeException):
+                candidate = AzureIPv4Range(ip_address_first, ip_address_last)
+                if not any(subnet.overlaps(candidate) for subnet in self.subnets):
+                    self.subnets.append(candidate)
+                    break
+            ip_address_first = ip_address_first + number_of_addresses
+        return candidate
```
This feels like an Advent of Code problem :smile:.

jemrobinson replied:
Yeah, sorry about that! If you think of a better alternative feel free to suggest one :)

JimMadge commented on data_safe_haven/pulumi/components/virtual_machine.py:
```diff
@@ -36,15 +36,17 @@ def __init__(
 
     @property
-    def os_profile(self) -> compute.OSProfileArgs:
+    def os_profile(self) -> compute.OSProfileArgs | None:
```
```suggestion
    def os_profile(self) -> Optional[compute.OSProfileArgs]:
```

JimMadge commented on data_safe_haven/pulumi/components/virtual_machine.py:
```diff
@@ -36,15 +36,17 @@ def __init__(
 
     @property
-    def image_reference(self) -> compute.ImageReferenceArgs:
+    def image_reference(self) -> compute.ImageReferenceArgs | None:
```
```suggestion
    def image_reference(self) -> Optional[compute.ImageReferenceArgs]:
```

JimMadge reviewed (changes requested):
This already got merged :open_mouth:.

Routing the Pulumi Inputs/Outputs and dealing with promises adds code, but I'm pretty confident it is a better solution overall. Should be more robust and it offloads handling state to Pulumi.

I don't like all the places that `typing.cast` has been introduced. It feels like sidestepping validating/sanitising input. Which will be necessary because we can't guarantee what a user or API will give to the program.
